internal/plugin/connectors/tcp/pg/startup.go

```diff
@@ -28,7 +28,8 @@ func (s *SingleUseConnector) Startup() error {
 
 	// Handle the case where the startup message was an SSL request.
 	if version == protocol.SSLRequestCode {
-		return fmt.Errorf("SSL not supported")
+		s.sslNotSupported()
+		return s.Startup()
```

Comment:
From the specification, when the client sends SSLRequest, the server must respond with `S` or `N`. If the response is `S`, SSL Handshake begins (this is not supported so we will never send `S`). In the case of the `N` response, the client must send StartupMessage, so it's the same start as `SSLMode=disable` so we need all the code in the Startup function again. 
So it's the same connection that will continue from the protocol start again.

Reply:
Thank you for the explanation. I can see the recursion resulting in an infinite loop since we rely on the client to provide the base case. One way we can circumvent this is by ensuring startup can at most recurse N (I assume once is enough) number of times before the connection closes.

Reply:
I agree, I'll fix the infinite loop. Is it a test that is failing? I run the bin/test and it looks fine :-/

Reply:
No, no failing test. Purely anticipation. It's definitely not a case that we test right now.

Reply:
👍  it's rewritten without a recursion now